tests/pytests/test_debug_commands.py

```diff
@@ -362,29 +366,285 @@ def testSpecIndexesInfo(env: Env):
+        expected_results_count = self.env.shardsCount * timeout_res_count
+        # set LIMIT to be larger than the expected results count
+        limit = expected_results_count + 1
+        self.QueryWithLimit(basic_debug_query, timeout_res_count, limit, expected_res_count=expected_results_count, should_timeout=True, message="SearchDebug:")
+
+        # SEARCH always has a sorter
+        self.QueryWithSorter()
+
+        # with no sorter (dialect 4)
+        self.QueryWithLimit(basic_debug_query, timeout_res_count, limit, expected_res_count=expected_results_count, should_timeout=True, message="SearchDebug:")
+
+        self.StrictPolicy()
+
+    def testSearchDebug(self):
+        self.SearchDebug()
+
+    def testSearchDebug_MT(self):
+        self.env.expect(config_cmd(), 'SET', 'WORKERS', 4).ok()
+        self.SearchDebug()
+        self.env.expect(config_cmd(), 'SET', 'WORKERS', 0).ok()
+
+    def AggregateDebug(self):
+        env = self.env
+        self.setBasicDebugQuery("AGGREGATE")
+        basic_debug_query = self.basic_debug_query
+        self.QueryDebug()
+
+        # EOF will be reached before the timeout counter
+        limit = 2
+        res = self.QueryWithLimit(basic_debug_query, timeout_res_count=10, limit=limit, expected_res_count=limit, should_timeout=False)
+
+        self.QueryWithSorter(sortby_params=['sortby', 1, '@n', 'load', 1, '@n'])
+
+        # with cursor
+        timeout_res_count = 200
+        limit = self.num_docs
+        cursor_count = 600 # higher than timeout_res_count, but lower than limit
+        debug_params = ["TIMEOUT_AFTER_N", timeout_res_count, "DEBUG_PARAMS_COUNT", 2]
+        cursor_query = [*basic_debug_query, 'WITHCURSOR', 'COUNT', cursor_count]
+        res, cursor = env.cmd(*cursor_query, 'LIMIT', 0, limit, *debug_params)
+        self.verifyResultsResp3(res, timeout_res_count, "AggregateDebug with cursor:")
+
+        iter = 0
+        total_returned = len(res['results'])
+        expected_results_per_iter = timeout_res_count
+
+        should_timeout = True
+        check_res_count = True
+        while (cursor):
+            # The counter of timeout simulator rp is initialized to timeout_res_count and decreases with each result returned from the upstream rp.
+            # In cluster mode, RPNet_next (the upstream rp) checks for timeouts at the end/start of reading each shard's reply.
+            # i.e we will always consume the entire chunk of results from each shard before checking for timeout.
+            # If some shards returns fewer than timeout_res_count results (due to EOF), we will continue to read replies until
+            # the sum of their results exceeds timeout_res_count.
+            # if the sum of results from all shards exceeds timeout_res_count but *each* shard individually returns fewer:
+            # no timeout warning is issued, the cursor is depleted.
```

Comment:
In this case, are we stopping whenever we reach an accumulated number of results >= timeout_res_count (even if we did not collect results from some shards)?

Reply:
didn't understand the question

Reply:
Refining and moving this comment to the aggregate debug header as it is relevant to the general behaviour of FT.AGGREGATE and not specifically to cursors.